### PR TITLE
Smart Contract First Version

### PR DESCRIPTION
**Pull request by rperez89**
Just a first draft of the smart contract to start reviewing it

The request function is the one that is going to create the vote.
The  _request function is the one that is going to be executed once the vote is approved.

I still need to do the claim function for the negative vote

### CODE REVIEW

File: contracts/TokenRequest.sol [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import "@aragon/os/contracts/lib/token/ERC20.sol";
+import "@aragon/os/contracts/lib/math/SafeMath.sol";
+import "./lib/ArrayUtils.sol";
+
+
+contract TokenRequest is AragonApp {
+    using SafeMath for uint256;
+
+    bytes32 constant public TOKEN_REQUEST_ROLE = keccak256("TOKEN_REQUEST_ROLE");
+    bytes32 constant public REFOUND_ROLE = keccak256("REFOUND_ROLE");
```
Comment (lkngtn) [inline contracts/TokenRequest.sol]: Should this be REFUND_ROLE?

**Comment by lkngtn** [comment]
Will leave the more thorough technical review to @willjgriff @0xGabi @rperez89 but did look through and excited to see this started. 🚀

For discussion, I wonder whether we want to have special handling for ETH or if we want to only accept ERC20s? (I expect that most people would want to do fundraising in something like DAI or USDC and not a volatile asset, and if they want to do ETH they could always do WETH).

On the flipside if we end up deploying to a sidechain like xDAI or an Aragon specific sidechain that uses a stable asset as the native fee token, perhaps having support for ETH as well makes sense. 

Curious what everyones thoughts are on 1) the added complexity of supporting ETH as well as tokens, and 2) thoughts on the marginal utility of supporting both. 

Personally was leaning towards just supporting tokens, but the prospect of deploying to something like xDAI makes me think we should support both.

File: contracts/TokenRequest.sol (hunk 1 above)
Comment (rperez89) [reply contracts/TokenRequest.sol]: uuppss  sure!

File: contracts/TokenRequest.sol [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        vault = _vault;
+        tokenManager = _tokenManager;
+    }
+
+    function request(
+        address _depositToken,
+        uint256 _depositAmount,
+        uint256 _mintAmount,
+        bytes _evmScript,
```
Comment (kanaagawa) [inline contracts/TokenRequest.sol]: Could we use `_executionScript` to follow the same standard as the rest of the apps?

File: contracts/TokenRequest.sol [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        require(request_.token = _depositToken, ERROR_DEPOSIT_TOKEN_MISMATCH);
+        require(request_.depositAmount = _depositAmount, ERROR_DEPOSIT_VALUE_MISMATCH);
+        require(request_.mintAmount = _mintAmount, ERROR_MINT_VALUE_MISMATCH);
+
+        request_.active = false;
+        if (_depositToken == ETH) {
+             vault.deposit.value(_depositAmount)(ETH, _depositAmount);
+        }
+
+        require(ERC20(_depositToken).safeApprove(vault, _depositAmount), ERROR_TOKEN_APPROVE_FAILED);
```
Comment (kanaagawa) [inline contracts/TokenRequest.sol]: Wouldn't line 129 and 130 should be in an `else` statement?

**Comment by 0xGabi** [comment]
@lkngtn As far as I know supporting ETH is fairly trivial so I belive we should include it. On the other hand, I might not be aware of the gas optimization considerations that could lean towards not implementing it

File: contracts/TokenRequest.sol (hunk 2 above)
Comment (rperez89) [reply contracts/TokenRequest.sol]: you mean the variable name? yes sure!

**Pull request merged**